torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -201,10 +200,79 @@ class ShapePropagator {
         continue;
       }
     }
-
     return tensor_types;
   }
 
+  c10::optional<c10::ScalarType> typeToScalarType(const c10::TypePtr & type) {
+    if(auto cast = type->cast<DimensionedTensorType>()) {
+      return cast->scalarType();
+    } else if (auto cast = type->cast<FloatType>()) {
+      return c10::ScalarType::Float;
+    } else if (auto cast = type->cast<IntType>()) {
+      return c10::ScalarType::Long;
+    } else if (auto cast = type->cast<BoolType>()) {
+      return c10::ScalarType::Bool;
+    }
+    return c10::nullopt;
+  }
+
+  // Promotes result types for arithmetic operations using new type promotion logic.
+  // See tensor_attributes.rst for details.
+  // This function handles the dimmed and zero-dim case, not the `wrapped-number`
+  // case. Wrapped number is handled separately because jit explicitly matches
+  // on the `Scalar other` signature.
+  c10::optional<c10::ScalarType> getPromotedTypeForArithmeticOp(Node *node) {
+    c10::ScalarType dimmed = c10::ScalarType::Undefined;
+    c10::ScalarType zerodim = c10::ScalarType::Undefined;
+    auto& args = node->schema().arguments();
+    // binary arithmetic ops, more than 2 args is alpha.
+    for (size_t i = 0 ; i < 2 ; i++ ) {
+      auto arg = args[i];
+      auto dtt = node->inputs()[i]->type()->cast<DimensionedTensorType>();
+      if ( dtt && dtt->dim() > 0) {
+        if (dtt->dim() > 0) {
+          if (dimmed == c10::ScalarType::Undefined) {
+            dimmed = dtt->scalarType();
+          } else {
+            dimmed = c10::promoteTypes(dimmed, dtt->scalarType());
+          }
+        }
+      } else if (!isFloatingType(dimmed)) {
+        // if not dimensioned
+        auto inputDtype = typeToScalarType(node->inputs()[i]->type());
```

Comment:
Since we're only iterating over the first two args, this should always be a dimensioned tensor type, so the `inputDtype ` will always be known. Also the `inputDtype` is used in both branches

Reply:
fixed